### ISSUE TITLE
Can I modify the  ElsaContext.ElsaSchema  externally？

### ISSUE THREAD

**Issue by netty2019** [elsa-workflows/elsa-core#2675]
When I migrate to Oracle, I always prompt that the user does not exist because the schema in Oracle is capitalized by default？
![image](https://user-images.githubusercontent.com/47128297/149879862-621d3be2-2b62-424e-80c7-f5cebe8f4562.png)
![image](https://user-images.githubusercontent.com/47128297/149880041-6d5b1456-335b-42bc-9bc1-0e41312e9afe.png)
![image](https://user-images.githubusercontent.com/47128297/149880119-05f6543a-650a-42ae-b10e-c0af6d842674.png)

**Comment by sfmskywalker** [comment]
We could probably allow the user to change the schema externally, but then you need to take care of generating your own migrations.

**Comment by netty2019** [comment]
Yes, it seems that the current migration only supports Oracle 12. It is incompatible with Oracle 11g. I regenerated the migration。
`    public class OracleContextFactory: IDesignTimeDbContextFactory<OracleContext>
    {
        public OracleContext CreateDbContext(string[] args)
        {
            var configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.Development.json", false)
                .Build();

            var dbContextBuilder = new DbContextOptionsBuilder<OracleContext>();
            dbContextBuilder.UseOracle<OracleContext>(configuration.GetConnectionString("Default"), b =>{ b.UseOracleSQLCompatibility("11"); });
            return new OracleContext(dbContextBuilder.Options);
        }
    }`

` public class OracleContext : ElsaContext
    {
        public const int MaxIdentifierLength = 30;

        protected string _schema = "HMCS";

        public override string Schema => _schema;

        public OracleContext(DbContextOptions<OracleContext> options) : base(options)
        {

        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            modelBuilder.HasDefaultSchema(_schema);
            foreach (IMutableEntityType entity in modelBuilder.Model.GetEntityTypes())
            {
                var tableName = entity.GetTableName();
                if (tableName == "Bookmarks" || tableName == "WorkflowDefinitions" || tableName== "WorkflowExecutionLogRecords" || tableName== "WorkflowInstances")
                {
                    //表名加前缀 转大写
                    entity.SetTableName("WF_" + tableName.ToUpper());
                    //列名转大写
                    foreach (var property in entity.GetProperties())
                    {
                        property.SetColumnName(property.Name.ToUpper());
                    }
                    //索引长度超长截取
                    foreach (var index in entity.GetIndexes())
                    {
                        var indexDBName = index.GetDatabaseName();

                        if (indexDBName.Length > MaxIdentifierLength) {
                            index.SetDatabaseName((indexDBName.Substring(0, MaxIdentifierLength-11)+"_"+"".GetUniqueShortString(10)).ToUpper());
                        }
                    }
                }
            }
            modelBuilder.Model.SetMaxIdentifierLength(30);
        }
    }`

**Comment by sfmskywalker** [comment]
OK, then let's allow for changing the schema name on the db context and add instructions that, when using this, one needs to take care of the migrations themselves (which should be documented).

**Comment by mabujahel** [comment]
Unhandled exception. System.InvalidOperationException: No suitable constructor was found for entity type 'WorkflowTrigger'. The following constructors had parameters that could not be bound to properties of the entity type: cannot bind 'workflowDefinitionId', 'activityId', 'activityType', 'bookmarkHash', 'bookmark', 'tenantId' in 'WorkflowTrigger(string workflowDefinitionId, string activityId, string activityType, string bookmarkHash, IBookmark bookmark, string tenantId)'Unhandled exception. System.InvalidOperationException: No suitable constructor was found for entity type 'WorkflowTrigger'. The following constructors had parameters that could not be bound to properties of the entity type: cannot bind 'workflowDefinitionId', 'activityId', 'activityType', 'bookmarkHash', 'bookmark', 'tenantId' in 'WorkflowTrigger(string workflowDefinitionId, string activityId, string activityType, string bookmarkHash, IBookmark bookmark, string tenantId)'

what about this problem ?
@sfmskywalker

**Comment by sfmskywalker** [comment]
Not sure, haven’t seen that. Do you have repro steps?

**Comment by mabujahel** [comment]
i am using a Oracle 11 database with a elsa 2.9 with dotnet core 6
i am create a new dbContext 
` public class ApplicationDbContext : ElsaContext
    {
        public const int MaxIdentifierLength = 30;

        protected string _schema = "DMS_ELSA";

        public override string Schema => _schema;

        public ApplicationDbContext(DbContextOptions<ApplicationDbContext> options) : base(options)
        {

        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            base.OnModelCreating(modelBuilder);
            modelBuilder.HasDefaultSchema(_schema);
            modelBuilder.ApplyConfigurationsFromAssembly(typeof(ApplicationDbContext).Assembly);
 

            foreach (IMutableEntityType entity in modelBuilder.Model.GetEntityTypes())
            {
                var tableName = entity.GetTableName();
                
                if (tableName != null && tableName.Length > 0)
                {
                    entity.SetTableName("WF_" + tableName.UseUpperSnakeCaseNamingConvention());

                    foreach (var property in entity.GetProperties())
                    {
                        property.SetColumnName(property.Name.UseUpperSnakeCaseNamingConvention());
                    }

                    foreach (var index in entity.GetIndexes())
                    {
                        var indexDBName = index.GetDatabaseName();

                        if (indexDBName.Length > MaxIdentifierLength)
                        {
                            index.SetDatabaseName((indexDBName.Substring(0, MaxIdentifierLength - 11) + "_" + "".GetUniqueShortString(10)).UseUpperSnakeCaseNamingConvention());
                        }
                    }
                }
            }

            modelBuilder.Model.SetMaxIdentifierLength(30);
        }


    }`
    and i am create a Factory 
    `public class ApplicationContextFactory : IDesignTimeDbContextFactory<ApplicationDbContext>
    {
        public ApplicationDbContext CreateDbContext(string[] args)
        {
            var configuration = new ConfigurationBuilder()
                .AddJsonFile("appsettings.json", false)
                .Build();

            var dbContextBuilder = new DbContextOptionsBuilder<ApplicationDbContext>();
            dbContextBuilder.UseOracle<ApplicationDbContext>(configuration.GetConnectionString("DefultConnection"), 
                b => { b.UseOracleSQLCompatibility("11"); })
             .UseUpperSnakeCaseNamingConvention(CultureInfo.CurrentCulture)
               ;

            return new ApplicationDbContext(dbContextBuilder.Options);
        }
    }`
    this is a costume function UseUpperSnakeCaseNamingConvention Convert Naming 
    `public static string UseUpperSnakeCaseNamingConvention(this string input)
        {
            string upperSnakeCase = Regex.Replace(input, "(?<!^)([A-Z][a-z]|(?<=[a-z])[A-Z])", "_$1").ToUpper();
            return upperSnakeCase;
        }`
        and in startup i am configuration elsa 
        ` services  
                .AddElsa(elsa => elsa
                .UseEntityFrameworkPersistence<ApplicationDbContext>(ef => ef.UseOracle(Configuration.GetConnectionString("DefultConnection")).UseUpperSnakeCaseNamingConvention(CultureInfo.CurrentCulture),false)
                 .AddConsoleActivities()
                   .AddHttpActivities(elsaSection.GetSection("Server").Bind) 
                   .AddWorkflowsFrom<Startup>()  
                   .AddActivitiesFrom<GetDocument>()
                   .AddActivity<CreateTask>() 
               );`

**Comment by mabujahel** [comment]
@sfmskywalker 
Application run If Delete Elsa Schema from database

 I am trying to connect to sql server database application run Perfectly

**Comment by mabujahel** [comment]
can give me sample url for elsa with oracle ?

**Comment by mabujahel** [comment]
can give me sample url for elsa with oracle ?
